libkbfs/coalescing_context.go

```diff
@@ -0,0 +1,110 @@
+package libkbfs
+
+import (
+	"golang.org/x/net/context"
+	"reflect"
+)
+
+// CoalescingContext allows many contexts to be treated as one.  It waits on
+// all its contexts' Context.Done() channels, and when all of them have
+// returned, this CoalescingContext is canceled. At any point, a context can be
+// added to the list, and will subsequently also be part of the wait condition.
+// TODO: add timeout channel in case there is a goroutine leak
+type CoalescingContext struct {
+	context.Context
+	closeCh  chan struct{}
```

Comment:
Well just save it in a `chan<- struct{}` field, rather than a full `chan struct{}` field (so you can close it in the loop).